### PR TITLE
feat: Add placeholder support for images

### LINKED ISSUE #172: Generate placeholders for `gatsbyImageData` resolver
Add `placeholder` argument to `gatsbyImageData` resolver and generate supported placeholders.

* Blurred image.

### PR DESCRIPTION
**Pull request by jpedroschmitz**
This PR adds support for a placeholder to the `gatsbyImageData` resolver. 🎉 

I've added 3 options for the placeholder, they are:

<details>
  <summary><code>BLURRED</code></summary>
  Generates a very low-resolution version of the image and displays it as a blurred background (default)
  
  blurred
</details>

<details>
  <summary><code>DOMINANT_COLOR</code></summary>
  The dominant color of the image used as a solid background color.
  
  dominantcolor
</details>

<details>
  <summary><code>TRACED_SVG</code></summary>
  Generates a simplified, flat SVG version of the source image, which it displays as a placeholder.

  tracedsvg
</details>

**Additional context:**

I think it's necessary to add this support under a beta tag, or alpha, so users can try it out and report any bugs before adding it as a stable feature.

Closes #172.

### CODE REVIEW

File: demo/package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "gatsby-plugin-image": "1.1.0",
     "gatsby-plugin-mdx": "2.1.0",
     "gatsby-plugin-postcss": "4.1.0",
+    "gatsby-plugin-sharp": "^3.1.2",
```
Comment (notrab) [inline demo/package.json:18]: We typically install with `-E` to pin the exact version. This is how Renovate is configured in this repo - so doing it here will mean no further PR is opened automatically.

File: gatsby-source-graphcms/package.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
   "license": "MIT",
   "dependencies": {
     "@babel/runtime": "7.12.5",
+    "gatsby-core-utils": "^2.1.0",
```
Comment (notrab) [inline gatsby-source-graphcms/package.json:20]: `yarn add gatsby-core-utils -E` here too

File: gatsby-source-graphcms/src/gatsby-node.js [hunk 3]
```diff
@@ -334,37 +341,85 @@ const generateImageSource = (
...
+function makeResolveGatsbyImageData(cache) {
+  return async function resolveGatsbyImageData(
+    { handle: filename, height, mimeType, width, url, internal },
+    options
+  ) {
+    const { placeholder = 'BLURRED' } = options
+
+    const imageDataArgs = {
+      ...options,
+      pluginName: `gatsby-source-graphcms`,
```
Comment (notrab) [inline gatsby-source-graphcms/src/gatsby-node.js:353]: Perhaps we can import `package.json` above, and use the `name` from there as that constant is already set.

```js
import pkg from '../package.json`
```

**Review by notrab — CHANGES_REQUESTED** [review]
@jpedroschmitz this is great!

I've left a few comments, but other than that, it should be good to go.

I'll change the base branch to `next` as I think that is configured for prereleases still.

File: gatsby-source-graphcms/src/gatsby-node.js (hunk 3 above)
Comment (jpedroschmitz) [reply gatsby-source-graphcms/src/gatsby-node.js]: I ran into issues while importing the JSON file, so I created a `constants` file on the util folder, and where I found the plugin name, I've refactored to use this constant. Let me know if this is okay.

**Review by notrab — APPROVED** [review]
Looks great @jpedroschmitz 🚀

**Comment by ynnoj** [comment]
@notrab Quick note, you'll need to update the branches configuration for [`semantic-release`](https://semantic-release.gitbook.io/semantic-release/usage/workflow-configuration#pre-release-branches) if your intention is to publish this as a prerelease.

**Comment by notrab** [comment]
@ynnoj sounds great! I'll get that added. Thanks!

**Comment by notrab** [comment]
Will merge. Hopefully the prerelease works!

**Review by notrab — APPROVED** [review]

**Comment by notrab** [comment]
@all-contributors please add @jpedroschmitz for code, example, ideas

**Comment by notrab** [comment]
@jpedroschmitz we're trying this out now on our website, and will see how it goes before we merge `next` into `main` here.

Thank you for such an awesome feat addition! 🚀 

PS. I've added you as a contributor to the package README 🥇

**Comment by jpedroschmitz** [comment]
That's awesome, thanks! Let me know if there's an issue with the placeholder feature =D

**Comment by notrab** [comment]
We had a report earlier from someone implementing this

> ERROR 
> [gatsby-source-graphcms] In order to use the traced svg placeholder, you need to install gatsby-plugin-sharp
>
> despite having
> "gatsby-plugin-sharp": "3.1.2"
> in the project dependencies

I haven't yet tried to replicate this @jpedroschmitz but will report back when I do.

**Comment by notrab** [comment]
Unable to duplicate. It must a local user error. Works fine for me in a new, and existing project with all placeholder types.

**Comment by jpedroschmitz** [comment]
That's weird. The plugin was added on the `gatsby-config` file?

If it was, maybe deleting the `node_modules` and `yarn.lock`, alongside with `gatsby clean`, can fix the issue.

**Pull request merged**